### DISCUSSION TITLE
xfixes.GetCursorImage how to

### DISCUSSION THREAD

**Discussion by codemodify** [jezek/xgb#26] [Q&A]
I'm trying to capture the screen with the cursor.
Seems I have to break this down into two steps
1. capture screen (I know how to do)
2. capture mouse cursor using xfixes.GetCursorImage() (I don't know how to unpack the output)

here is a sample code
```
	var cursorImageCookie xfixes.GetCursorImageCookie
	cursorImageCookie = xfixes.GetCursorImage(X)

	var cursorImageReply *xfixes.GetCursorImageReply
	cursorImageReply, err = cursorImageCookie.Reply()
	if err != nil {
		fmt.Println(err)
		os.Exit(-1)
	}

	cursorImageReply.CursorImage // what to do with this

```

how do I convert the `cursorImageReply.CursorImage` to a golang `image.Image` or any other golang image formats

**Comment by codemodify** [comment]
Also when calling the function, I'm getting this

![image](https://github.com/jezek/xgb/assets/44710208/e71d40f4-8b39-431f-afc6-8c282046c13a)

**Comment by jezek** [comment]
> how do I convert the cursorImageReply.CursorImage to a golang image.Image or any other golang image formats

I've never worked, with `xfixes` extension, but whenI read the docs of [type GetCursorImageReply](https://github.com/burntsushi/xgb/blob/deaf085860bc/xfixes/xfixes.go#L1388), there is a comment after 
CursorImage `// size: xgb.Pad(((int(Width) * int(Height)) * 4))`. Which could indicate there is a image in RGBA format stored there byte by byte. Basicaly xgb (x bindings for go) is the same as xcb (x bindings for c), so if I was you, I would search in xcb documentation, how the image is stored.

> Also when calling the function, I'm getting this

Please don't attach images with error messages. Always copy-paste text. It's easier to work with. Also the whole error message is very badly readable from your image.

**Comment by codemodify** [comment]
sorry, this is the sample code, the line `cursorImageReply, err = cursorImageCookie.Reply()` returns `BadRequest {NiceName: Request, Sequence: 2, BadValue: 0, MinorOpcode: 4, MajorOpcode: 138}`

```go
func main() {
	X, err := xgb.NewConn()
	if err != nil {
		fmt.Println(err)
		return
	}
	defer X.Close()

	// xproto.Setup retrieves the Setup information from the setup bytes
	// gathered during connection.
	setup := xproto.Setup(X)

	// This is the default screen with all its associated info.
	screen := setup.DefaultScreen(X)
	fmt.Println(screen.Root)

	err = xfixes.Init(X)
	if err != nil {
		fmt.Println(err)
		os.Exit(-1)
	}

	var cursorImageCookie xfixes.GetCursorImageCookie
	cursorImageCookie = xfixes.GetCursorImage(X)

	var cursorImageReply *xfixes.GetCursorImageReply
	cursorImageReply, err = cursorImageCookie.Reply()
	if err != nil {
		fmt.Println(err)
		os.Exit(-1)
	}

	fmt.Println(cursorImageReply.CursorImage)
}

```

**Comment by aarzilli** [comment]
You are not initializing the extension correctly: https://www.x.org/releases/X11R7.7/doc/fixesproto/fixesproto.txt

**Comment by codemodify** [comment]
the library api only has this `err = xfixes.Init(X)` to init, does it mean the implementation is not up to date ?

**Comment by jezek** [comment]
I've asked [ChatGPT](https://chatgpt.com/share/a94bcdfb-34b0-43b5-959a-9fac2847c404) how to do it and it suggests querying for version after xfixes init (for both xcb and xgb version) like this:
```go
    // Query xfixes version
    versionCookie := xfixes.QueryVersion(conn, xfixes.MajorVersion, xfixes.MinorVersion)
    versionReply, err := versionCookie.Reply()
    if err != nil {
        log.Fatalf("Error: Cannot get xfixes version: %v\n", err)
    }
```

Note: I haven't tested it, so I don't know if ChatGPT is right, but from my previous experience it is mostly right and a good helper for debuging and source for inspiration.

Note2: If you succeed to get and decode the cursor image, could you write an example (into [examples](https://github.com/jezek/xgb/tree/master/examples) dir) for getting the image and let's say writing it to a file, please? It would be a big help for future package users.

**Comment by codemodify** [comment]
ok so I figured it out, but using raw xorg apis, will have to convert it to xgb analogy

one more questions maybe you know the answer
how can I query for the current mouse/cursor shape ?
ex, the function either Go or C would return: 
- if current focus is on a text field then it would be "MouseText"
- if current focus is generic then it would be "MouseArrow"
- if current focus is resizing a window then it would be "MouseResize"

etc

**Comment by jezek** [comment]
I don't know answer for your question, but my guess is, that xgb doesn't know anything about text fields in app, so the app has to set the cursors by itself. Just for curiosity I've [asked ChatGPT](https://chatgpt.com/share/68de726d-28ef-4242-b889-c25f48760d8c) your question, hope it helps.

On the other note. I think this is not an issue. I'll convert this thread to Discussions. You can set this thread as answered, if you found the solution (feel free to share your findings) and maybe open new for your next question above.

Have a nice time.
